tests/js/couchConfigTests.js

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2017 OCAD University
+Licensed under the Educational Community License (ECL), Version 2.0 or the New
+BSD license. You may not use this file except in compliance with one these
+Licenses.
+You may obtain a copy of the ECL 2.0 License and BSD License at
+https://raw.githubusercontent.com/fluid-project/fluid-couch-config/master/LICENSE.txt
+*/
+
+/* global emit, sum  */
+
+"use strict";
+
+var fluid = require("infusion");
+
+var gpii  = fluid.registerNamespace("gpii");
+
+require("gpii-pouchdb");
+gpii.pouch.loadTestingSupport();
+
+var jqUnit = require("node-jqunit");
+
+require("../../src/couchConfig");
+
+"use strict";
+
+fluid.defaults("fluid.tests.couchConfig.testCouchConfig", {
+    gradeNames: ["fluid.couchConfig.db", "fluid.couchConfig.documents", "fluid.couchConfig.designDocument"],
+    dbConfig: {
+        couchURL: "http://localhost:6789",
+        dbName: "testDbForTests",
+        designDocName: "testViews"
+    },
+    dbDocuments : {
+        testDoc: {
+            "type": "test",
+            "key": "value",
+            "arrayKey": ["values", "in", "an", "array"]
+        },
+        testDoc2: {
+            "type": "test",
+            "key": "value2",
+            "arrayKey": ["values", "in", "an", "array"]
+        }
+    },
+    dbViews: {
+        test: {
+            map: "fluid.tests.couchConfig.testMapFunction",
+            reduce: "fluid.tests.couchConfig.testReduceFunction"
+        }
+    },
+    dbValidate: {
+        validateFunction: "fluid.tests.couchConfig.testValidateFunction"
+    }
+});
+
+fluid.defaults("fluid.tests.couchConfig.couchConfigTester", {
```

Comment:
I believe that we are missing some code branches in our testing:

- Update a document when it already exists and is the same (and verify that we didn't update it anyway)
- Update a document when it already exists and is different
- Update a design document when it already exists and is the same
- Update a design document when it already exists and is different
- Each of the different function types supported in `fluid.couchConfig.designDocument.generateViews()`

It would also be worthwhile testing when a database already exists with documents not in our configuration and verifying that they are left alone.

And once we have programmatic reporting of error conditions, testing error conditions.

Reply:
To note: `fluid.couchConfig.designDocument.generateViews()` has been renamed and refactored into `fluid.couchConfig.updateDesignDocument.renderViewFunctions()` and `fluid.couchConfig.updateDesignDocument.renderFunctionString()`